cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java

```diff
@@ -0,0 +1,143 @@
+
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import io.cdap.cdap.app.runtime.ProgramRuntimeService;
+import io.cdap.cdap.common.NamespaceNotFoundException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.namespace.NamespacePathLocator;
+import io.cdap.cdap.common.namespace.NamespaceQueryAdmin;
+import io.cdap.cdap.gateway.handlers.util.AbstractAppFabricHttpHandler;
+import io.cdap.cdap.internal.app.services.ApplicationLifecycleService;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.http.HttpHandler;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+
+import java.io.File;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+
+/**
+ * Internal {@link HttpHandler} for Application Lifecycle Management
+ */
+@Singleton
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3 + "/namespaces/{namespace-id}")
+public class AppLifecycleHttpHandlerInternal extends AbstractAppFabricHttpHandler {
+  private static final Gson GSON = new Gson();
+
+  /**
+   * Runtime program service for running and managing programs.
+   */
+  private final ProgramRuntimeService runtimeService;
+
+  private final CConfiguration configuration;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final NamespacePathLocator namespacePathLocator;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final File tmpDir;
+
+  @Inject
+  AppLifecycleHttpHandlerInternal(CConfiguration configuration,
+                                  ProgramRuntimeService runtimeService,
+                                  NamespaceQueryAdmin namespaceQueryAdmin,
+                                  NamespacePathLocator namespacePathLocator,
+                                  ApplicationLifecycleService applicationLifecycleService) {
+    this.configuration = configuration;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.runtimeService = runtimeService;
+    this.namespacePathLocator = namespacePathLocator;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.tmpDir = new File(new File(configuration.get(Constants.CFG_LOCAL_DATA_DIR)),
+                           configuration.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+  }
+
+  /**
+   * Get a list of {@link ApplicationDetail} for all applications in the given namespace
+   *
+   * @param request   {@link HttpRequest}
+   * @param responder {@link HttpResponse}
+   * @param namespace the namespace to get all application details
+   * @throws Exception if namespace doesn't exists or failed to get all application details
+   */
+  @GET
+  @Path("/appdetail")
```

Comment:
nit: collections should be plural -> 'appdetails'. Though I think it's better to just have it as 'apps'

Reply:
Fixed. Changed to /apps and /app/<appid>